File: Queries with bags/Solution/queries_with_bags.cpp
/// ILLUMINATI SEASON 6
/// Queries with Bags

#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;

#define FAST_IO ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

const ll MAX = 1e5+1;
vector<ll> parent(MAX);
vector<ll> unionBySize(MAX,1);

ll find_set(ll v){
    if(parent[v]==v)return v;
    return parent[v] = find_set(parent[v]);
}

void union_set(ll a, ll b){
    if(unionBySize[a] < unionBySize[b])swap(a,b);
    parent[b] = a;
    unionBySize[a] += unionBySize[b];
}

int main(){
    FAST_IO
    ll n,q;
    cin>>n>>q;
    for(ll i=1;i<=n;i++)parent[i]=i;
    ll a, b;
    ll no_of_bags = n;
    vector<string> queries;
    while(q--){
        ll query_type;
        cin>>query_type;
        if(query_type==1){
            cin>>a>>b;
            a = find_set(a);
            b = find_set(b);
            if(a!=b){
                no_of_bags--;
                union_set(a,b);
            }
        }
        else if(query_type==2){
            cin>>a>>b;
            a = find_set(a);
            b = find_set(b);

            if(a==b) queries.push_back("YES");
            else queries.push_back("NO");
        }
        else{
    		cout<<no_of_bags<<endl;
            string s = "";
            s+=x;
            queries.push_back(s);
        }
    }
    
}

File: Queries with bags/Solution/queries_with_bags.py
#ILLUMINATI SEASON 6
'''
Name : Jaymeet Mehta
codechef id :mj_13
Problem : Queries with bags
'''
from sys import stdin,stdout

def find_set(v):
    global parent
    if parent[v]!=v:
        parent[v]= find_set(parent[v])
    return parent[v]
def union_set(a,b):
    global size,parent
    if size[a]<size[b]:
        a,b=b,a
    parent[b]=a
    size[a]+=size[b]
N,Q = map(int,stdin.readline().split())
assert(1<=N and N<=100000)
assert(1<=Q and Q<=100000)
parent=[i for i in range(N)]
size=[1 for i in range(N)]
bags=N
for _ in range(Q):
    query= [int(x) for x in stdin.readline().split()]
    if query[0]==1:
        assert(query[1]!=query[2])
        assert(1<=query[1] and query[1]<=N)
        assert(1<=query[2] and query[2]<=N)
        a=find_set(query[1]-1)
        b=find_set(query[2]-1)
        if a!=b:
            bags-=1
            union_set(a,b)
        continue
    if query[0]==2:
        assert(query[1]!=query[2])
        assert(1<=query[1] and query[1]<=N)
        assert(1<=query[2] and query[2]<=N)
        a=find_set(query[1]-1)
        b=find_set(query[2]-1)
        if a!=b:
            print("NO")
        else:
            print("YES")
        continue
    print(bags)